### ISSUE TITLE
Potential StackOverflow & Wasted CPU Cycles

### ISSUE THREAD

**Issue by speige** [AnderssonPeter/CompressedStaticFiles#9]
In file https://github.com/AnderssonPeter/CompressedStaticFiles/blob/master/src/CompressedStaticFiles/CompressedStaticFileMiddleware.cs

These lines have a potential issue:
```
 var originalPrepareResponse = staticFileOptions.Value.OnPrepareResponse;
            staticFileOptions.Value.OnPrepareResponse = ctx =>
            {
                originalPrepareResponse(ctx);
```

The above code essentially "monkey-patches" the OnPrepareResponse. By keeping a reference to the original & adding some extra code.

The problem with this approach is that .net core instantiates a new Middleware instance on each request. However, the options.value passed to the middleware constructor is usually a singleton (depends on the programmer).

This means the 1st request will work fine, but the 2nd request will monkey-patch the already monkey-patched method. The double-monkey-patched method still works, it just wastes cpu time running the extra code 2x. With every request this overhead gets worse until a stack overflow exception eventually happens and the app pool resets.

One potential solution would be to have the constructor to CompressedStaticFileMiddleware.cs clone the original StaticFileOptions before monkey-patching.

**Comment by speige** [comment]
I believe a workaround would be to call 
```
app.Use(async (context, next) => ...)
```
instead of 
```
app.UseCompressedStaticFiles( ... )
```

and always instantiate a new instance of the Options parameter.

**Comment by Tratcher** [comment]
No, middleware are not recreated per request, only mvc does that with controllers.
Did you actually experience this issue or was it just a theory?

**Comment by Tratcher** [comment]
Ah, that Use pattern would recreate the middleware per request, it executes that func every request. UseMiddleware avoids this.
